safekeeper/src/send_wal.rs

```diff
@@ -633,6 +643,50 @@ impl<IO: AsyncRead + AsyncWrite + Unpin> WalSender<'_, IO> {
                 .await?;
         }
     }
+
+    /// Wait until we have available WAL > start_pos or timeout expires. Returns
+    /// - Ok(Some(end_pos)) if needed lsn is successfully observed;
+    /// - Ok(None) if timeout expired;
+    /// - Err in case of error -- only if 1) term changed while fetching in recovery
+    ///   mode 2) watch channel closed, which must never happen.
+    async fn wait_for_lsn(&mut self) -> anyhow::Result<Option<Lsn>> {
+        fail::fail_point!(
+            "sk-pause-send",
+            self.appname.as_deref() != Some("pageserver"),
+            |_| {
+                tokio::time::sleep(POLL_STATE_TIMEOUT);
```

Comment:
This is broken because of no .await. Again, have a look at failpoint_support::sleep_millis_async , and it is actually better to use it here.

Reply:
Yes sorry I have that fixed in my local changes - I want to just put the fix in this PR if that's ok

Reply:
Also why is `failpoint_support::sleep_millis_async` better here? It seems that relies on you configuring the failpoint to sleep, meaning the timeout is configurable by the failpoint. This would require the user to iknow POLL_STATE_TIMEOUT when configuring the endpoint, which seems wrong